Fix #416: Optimize ensemble prediction performance by reusing sparse vectors

mattsgarlata opened the pull request:
## Summary

This PR optimizes prediction performance for tree-based ensemble models by reusing sparse vector representations across ensemble members.

## Changes

1. **Added TreeEnsembleModel**: A new optimized ensemble model specifically for tree-based models
   - Overrides predict() to create the sparse vector once and reuse it across all trees
   - Falls back to standard prediction for non-tree models in the ensemble

2. **Enhanced TreeModel**: Added predict(SparseVector, Example) overload
   - Allows predictions using pre-computed sparse vectors
   - Avoids redundant sparse vector creation in ensemble scenarios

3. **Updated IndependentRegressionTreeModel**: Implemented the sparse vector overload
   - Consistent with TreeModel optimization

4. **Modified RandomForestTrainer and ExtraTreesTrainer**: 
   - Updated createEnsemble() to return TreeEnsembleModel instead of WeightedEnsembleModel
   - Provides automatic optimization for tree ensemble predictions

5. **Updated BaggingTrainer**: Made createEnsemble() method protected
   - Allows subclasses to customize ensemble model creation

6. **Added comprehensive tests**: TreeEnsembleOptimizationTest validates:
   - TreeEnsembleModel creation by trainers
   - Equivalence of sparse vector and standard prediction methods
   - Correctness of optimized ensemble predictions

## Performance Impact

This optimization eliminates redundant sparse vector creation when an ensemble of N trees makes predictions. For large ensembles, this can provide significant performance improvements, especially when examples have many features.

## Backward Compatibility

All changes are backward compatible:
- New methods/classes don't break existing APIs
- Existing models continue to work unchanged
- Tests verify functional equivalence

Fixes #416

Craigacp commented on Core/src/main/java/org/tribuo/ensemble/BaggingTrainer.java:
```diff
@@ -124,7 +130,7 @@ public BaggingTrainer(Trainer<T> trainer, EnsembleCombiner<T> combiner, int numM
      * @param numMembers The number of ensemble members to train.
      * @param seed The RNG seed used to bootstrap the datasets.
-     * @param numThreads The number of threads to use for parallel training. Set to -1 to use all available processors, 1 for single-threaded.
+     * @param numThreads The number of threads to use for parallel training. Set to USE_ALL_AVAILABLE_PROCESSORS to use all available processors, 1 for single-threaded.
```
This should be a link to the field `{@link #USE_ALL_AVAILABLE_PROCESSORS}`.

Craigacp commented on Common/Trees/src/main/java/org/tribuo/common/tree/ExtraTreesTrainer.java:
```diff
@@ -75,6 +81,22 @@ public ExtraTreesTrainer(DecisionTreeTrainer<T> trainer, EnsembleCombiner<T> com
+     * @param numMembers The number of ensemble members to train.
+     * @param seed The RNG seed.
+     * @param numThreads The number of threads to use for parallel training.
+     *                   Use USE_ALL_AVAILABLE_PROCESSORS to use all available processors, 1 for single-threaded.
```
This should be a link to the field `{@link BaggingTrainer#USE_ALL_AVAILABLE_PROCESSORS}`.

mattsgarlata replied:
Addressed in 3d5ba968a33002860fb05baa9a9d560731b804cb

mattsgarlata commented on Common/Trees/src/main/java/org/tribuo/common/tree/TreeEnsembleModel.java:
```diff
@@ -0,0 +1,95 @@
+ * Optimized ensemble model for tree-based models.
+ * <p>
+ * This subclass of {@link WeightedEnsembleModel} provides optimized prediction performance
+ * for ensembles containing {@link TreeModel}s. Non-tree models in the ensemble fall back
```
Addressed in 20d63f32890492b0cce5f6cc3250184418b5200c

mattsgarlata commented:
@Craigacp I think I addressed all your comments. Most were pretty straightforward, but you'll probably want to take a close look at the changes for https://github.com/oracle/tribuo/pull/417#discussion_r2538368990 which are in https://github.com/oracle/tribuo/commit/e170db38a7e4a6505f2f78978cda90468285d33a
